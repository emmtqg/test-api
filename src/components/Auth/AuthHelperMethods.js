import decode from 'jwt-decode';

import config from '../Config';

// The apiCamelizedLookup provides a fieldname check and crossreference
// for the form field names. It aids in translation of valid property
// names to whatever the API required (single point of definition between
// the state properties and api names). It was assumed if the ESLint rules
// did not allow underscores (as specified in the API names) then, rather
// than overriding the linter, a mechanism should be incorporated to allow
// any API to specify properties as required and they can then be updated
// in one location. If form fields are checked for testing, then the form
// component can also utilize the apiCamelizedLookup for field specifications.
const { apiCamelizedLookup } = config;
const apiNames = Object.keys(apiCamelizedLookup);

export default class AuthHelperMethods {
  getApiPropName = (name) => {
    // Get the API 'expected' property of input form field
    let index = -1;
    Object.values(apiCamelizedLookup).findIndex((ele, i) => {
      index = i;
      return (ele === name);
    });
    return index; // returning matched index or -1 if not in apiCamelized object
  }

  lookupApiName = (name) => {
    const index = this.getApiPropName(name);
    if (index > -1) {
      // apiNames are the API keys in lookup array
      return apiNames[index];
    }

    return ''; // API will error with invalid fieldname
  }

  register = (registerObject) => {
    const {
      firstName,
      lastName,
      email,
      password,
      confirmPassword,
    } = registerObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.registerUrl, {
        method: 'POST',
        body: JSON.stringify({
          [this.lookupApiName('firstName')]: firstName,
          [this.lookupApiName('lastName')]: lastName,
          email,
          password,
          [this.lookupApiName('confirmPassword')]: confirmPassword,
        }),
      })
        .then((response) => {
          // console.log(`Auth (Register): ${response}`);
          if (response) {
            this.setToken(response.token); // Setting the token in localStorage
            return Promise.resolve(response);
          }
          return Promise.reject(response);
        })
    );
  }

  newPlayer = (newPlayerObject) => {
    const {
      firstName,
      lastName,
      rating,
      handedness,
    } = newPlayerObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.playersUrl, {
        method: 'POST',
        body: JSON.stringify({
          [this.lookupApiName('firstName')]: firstName,
          [this.lookupApiName('lastName')]: lastName,
          rating,
          handedness,
        }),
      })
    );
  }

  getPlayers = () => (
    this.fetch(config.playersUrl, {
      method: 'GET',
    })
  );

  removePlayer = (playerId) => {
    const deleteUrl = `${config.playersUrl}/${playerId}`;
    return (
      this.fetch(deleteUrl, {
        method: 'DELETE',
      })
    );
  }

  login = (loginObject) => {
    const {
      email,
      password,
    } = loginObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.loginUrl, {
        method: 'POST',
        body: JSON.stringify({
          email,
          password,
        }),
      })
        .then((response) => {
          if (response) {
            // console.log(`Auth (Login): ${response}`);
            this.setToken(response.token); // Setting the token in localStorage
            return response;
          }
          return Promise.reject(response);
        })
        .catch((error) => {
          if (error && error.message) {
            // console.log(`Error: ${error.message}`);
          }
          return Promise.reject(error);
        })
    );
  }

  loggedIn = () => {
    // Checks if there is a saved token and it's still valid
    const token = this.getToken(); // Getting token from localstorage
    return (typeof token !== typeof undefined) && (!!token && !this.isTokenExpired(token));
  }

  isTokenExpired = (token) => {
    try {
      const decoded = decode(token);
      if (decoded.exp < Date.now() / 1000) { // Checking if token is expired.
        return true;
      }
      return false;
    } catch (err) {
      // console.log('Check for expired tolen failed.');
      return false;
    }
  }

  setToken = (idToken) => {
    // Saves user token to localStorage
    localStorage.setItem('id_token', idToken);
  }

  getToken = () =>
    // Retrieves the user token from localStorage
    localStorage.getItem('id_token')


  logout = () => {
    // Clear user token and profile data from localStorage
    localStorage.removeItem('id_token');
  }

  getConfirm = () => {
    // Using jwt-decode npm package to decode the token
    const answer = decode(this.getToken());
    // console.log(`Get confirm token => ${answer}`);
    return answer;
  }

  fetch = (url, options) => {
    // performs api calls sending the required authentication headers
    const headers = {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    };

    // Setting Authorization header
    // Authorization: Bearer xxxxxxx.xxxxxxxx.xxxxxx
    if (this.loggedIn()) {
      headers.Authorization = `Bearer ${this.getToken()}`;
    }

    return fetch(url, {
      headers,
      ...options,
    })
      .then(this.checkStatus)
      .then(response => response.json())
      .catch((error) => {
        // console.log(`Error: ${error.message}`);
        throw error;
      });
  }

  checkStatus = (response) => {
    // Success status lies between 200 to 300
    if (response.status >= 200 && response.status < 300) {
      return response;
    }

    // raises an error in case response status is not a success
    const error = new Error(response.statusText);
    error.response = response;
    error.status = response.status;
    throw error;
  }
}
