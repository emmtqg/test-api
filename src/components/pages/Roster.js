import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import Fab from '@material-ui/core/Fab';
import AddIcon from '@material-ui/icons/Add';
import { withStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import Paper from '@material-ui/core/Paper';
import AlertDialog from './AlertDialog';

import RosterPlayer from './RosterPlayer';
import '../styles/SideBySideComponents.scss';

// Layout styles
const styles = theme => ({
  root: {
    minWidth: 500,
    maxWidth: 720,
    flexGrow: 1,
    marginLeft: 'auto',
    marginRight: 'auto',
    overflowX: 'auto',
    fontSize: '1.6rem',
  },
  main: {
    width: 'auto',
    display: 'block', // Fix IE 11 issue.
    marginTop: 50,
    [theme.breakpoints.down('sm')]: {
      marginLeft: 'auto',
      marginRight: 'auto',
    },
    [theme.breakpoints.up((400 + theme.spacing.unit) * 3 * 2)]: {
      marginLeft: 'auto',
      marginRight: 'auto',
    },
  },
  paper: {
    minWidth: 400,
    maxWidth: 700,
    marginLeft: 'auto',
    marginRight: 'auto',
    marginTop: theme.spacing.unit * 8,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    padding: '.3rem',
  },
  table: {
    minWidth: 400,
    maxWidth: 700,
    marginTop: 10,
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  tableCell: {
    fontSize: '1.6rem',
  },
  fab: {
    margin: theme.spacing.unit,
    color: 'green',
  },
  extendedIcon: {
    marginRight: theme.spacing.unit,
  },
  alignTop: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  col: {
    display: 'inline-block',
    margin: '10 10',
  },
  subTitle: {
    marginBottom: '2rem',
  },
});

class Roster extends Component {
  constructor(props) {
    super(props);
    this.state = {
      players: [],
      message: '',
      dialogShow: false,
      dialogMessage: '',
      yesButtonTxt: 'OK',
      noButtonTxt: '',
      removeId: '',
      cbHandleConfirm: () => {},
    };

    this.addPlayer = this.addPlayer.bind(this);
    this.removePlayer = this.removePlayer.bind(this);
    this.showDialog = this.showDialog.bind(this);
    this.hideDialog = this.hideDialog.bind(this);
  }

  componentWillMount() {
    if (!this.props.auth.loggedIn()) {
      this.props.history.replace('/login');
    }
  }

  componentDidMount() {
    this.props.auth.getPlayers()
      .then((response) => {
        if (response) {
          const { players } = response;
          const camelizedPlayers = players.map(p =>
            ({
              firstName: p.first_name,
              lastName: p.last_name,
              id: p.id,
              rating: p.rating,
              handedness: p.handedness,
            }));

          this.setState({
            players: camelizedPlayers,
          });

          this.setPlayerCount();
        }
      })
      .catch((error) => {
        this.setState({
          players: [],
          message: error.message,
        });
      });
  }

  setPlayerCount() {
    const nPlayers = this.state.players.length;
    const getPlayersString = (n) => {
      const nplayerString = {
        0: 'are no players',
        1: 'is 1 player',
        default: `are ${nPlayers} players`,
      };
      return (nplayerString[n] || nplayerString.default);
    };
    const playersTxt = getPlayersString(nPlayers);
    this.setState({
      message: `There ${playersTxt} on your roster.`,
    });
  }

  getPlayerById(id) {
    return this.state.players.find(p => p.id === id);
  }

  addPlayer() {
    this.props.history.replace('/player/new');
  }

  showDialog = (id) => {
    let message = '';
    const player = this.getPlayerById(id);
    if (typeof player !== typeof undefined) {
      message = `Please confirm removal of ${player.firstName} ${player.lastName}.`;
      this.setState({
        removeId: id,
        dialogShow: true,
        dialogMessage: message,
        yesButtonTxt: 'Remove',
        noButtonTxt: 'Cancel',
        cbHandleConfirm: this.removePlayer,
      });
    } else {
      message = 'There was a problem removing this player: their ID was not recognized on the current roster.';
      this.setState({
        removeId: id,
        dialogShow: true,
        dialogMessage: message,
        yesButtonTxt: 'OK',
        noButtonTxt: '',
        cbHandleConfirm: this.hideDialog,
      });
    }
  };

  hideDialog = () => {
    this.setState({
      removeId: '',
      dialogShow: false,
      dialogMessage: '',
      yesButtonTxt: 'OK',
      noButtonTxt: '',
      cbHandleConfirm: this.removePlayer,
    });
  };

  removePlayer() {
    this.hideDialog();
    const { removeId } = this.state;
    this.props.auth.removePlayer(removeId)
      .then((response) => {
        if (response.success && response.success === true) {
          this.setState({
            players: this.state.players.filter(player => (player.id !== removeId)),
            removeId: '',
          });
          this.setPlayerCount();
        }
      })
      .catch((error) => {
        this.setState({
          removeId: '',
          dialogShow: true,
          dialogMessage: error.message,
          cbHandleConfirm: this.hideDialog,
          yesButtonTxt: 'OK',
          noButtonTxt: '',
        });
      });
  }

  render() {
    const { classes, ...other } = this.props;
    const { message } = this.state;

    return (
      <main className={classes.main}>
        <div id="Roster">
          <AlertDialog
            show={this.state.dialogShow}
            yesButtonTxt={this.state.yesButtonTxt}
            noButtonTxt={this.state.noButtonTxt}
            cbHandleConfirm={this.state.cbHandleConfirm}
            cbHandleClose={this.hideDialog}
            message={this.state.dialogMessage}
          />
          <div classes={classes.alignTop}>
            <div className={classes.col}>
              <Typography component="h5" variant="h5">
                <em>The</em> Roster
              </Typography>
            </div>
            <div className={classes.col}>
              <Fab size="small" aria-label="Add" className={classes.fab}>
                <AddIcon onClick={this.addPlayer} />
              </Fab>
            </div>
          </div>
          <Typography component="h6" variant="h6" className={classes.subTitle}>
            {message}
          </Typography>
          <Paper className={classes.root}>
            <Table className={classes.table}>
              <TableHead>
                <TableRow>
                  <TableCell className={classes.tableCell}>Player</TableCell>
                  <TableCell className={classes.tableCell} align="right">Rating</TableCell>
                  <TableCell className={classes.tableCell} align="right">Preferred Hand</TableCell>
                  <TableCell className={classes.tableCell} align="center">Remove</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {this.state.players.map(player => (
                  <RosterPlayer
                    {...other}
                    key={player.id}
                    player={player}
                    cbRemovePlayer={this.showDialog}
                      // this.removePlayer(player.id)}
                  />
                ))}
              </TableBody>
            </Table>
          </Paper>
        </div>
      </main>
    );
  }
}

Roster.defaultProps = {
  history: () => {
    this.setState({
      dialogShow: true,
      dialogMessage: 'No history object passed to this page [Roster]',
    });
  },
};

Roster.defaultProps = {
  auth: {},
};

Roster.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  classes: PropTypes.object.isRequired,
  history: PropTypes.shape({
    replace: PropTypes.func,
  }),
  auth: PropTypes.shape({
    loggedIn: PropTypes.func.isRequired,
    fetch: PropTypes.func.isRequired,
    getPlayers: PropTypes.func.isRequired,
    removePlayer: PropTypes.func.isRequired,
  }),
};

export default withStyles(styles)(Roster);
