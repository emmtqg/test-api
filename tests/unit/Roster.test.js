import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import Roster from '../../src/components/Roster';
import AuthHelperMethods from '../../src/components/Auth/AuthHelperMethods';

let UnwrappedShallowComponent;
let wrapper;

const auth = new AuthHelperMethods();

// auth mocks
auth.loggedIn = jest.fn(() => true); // replace loggedIn test for component mount check
auth.setToken = jest.fn(() => {
  auth.setToken('eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiI1YzlhNzFiMDMxMGQyMTdmZjM5MDFlMjkiLCJpYXQiOjE1NTM3MTQzNTd9.sfyRdYlfXXzgqFfxLuEONxh1TMG1teWTAMO_GvG4Xdw');
});

// auth mocks for Promise return of getPlayers API call from AuthHelpers.js
const playerList = {
  success: true,
  players: [{
    first_name: 'liam',
    last_name: 'neeson',
    rating: 9000,
    handedness: 'left',
    id: '5c9fdeee310d217ff3901e67',
  }, {
    first_name: 'U',
    last_name: '2',
    rating: 9000,
    handedness: 'right',
    id: '5ca0eb2f310d217ff3901e71',
  }, {
    first_name: 'kevin',
    last_name: 'bacon',
    rating: 3000,
    handedness: 'right',
    id: '5ca0eb50310d217ff3901e72',
  }, {
    first_name: 'sam2',
    last_name: 'magoooo',
    rating: 9000,
    handedness: 'left',
    id: '5ca771be310d217ff3901ead',
  }],
};
const rosterMock = jest.fn().mockReturnValue(Promise.resolve(playerList));
auth.getPlayers = rosterMock;

// history mock
const history = {
  replace: jest.fn(),
};

beforeAll(() => {
  wrapper = mount(<Roster auth={auth} props={history} />);
  const shallowComponent = shallow(<Roster auth={auth} props={history} />);
  UnwrappedShallowComponent = shallowComponent.first().shallow();
});

afterAll(() => {
  wrapper.unmount();
});

describe('The Roster Table should', () => {
  it('match the Roster form snapshot', () => {
    expect(wrapper.find('Roster').length).toEqual(1);
  });

  it('render correctly', () => {
    const component = renderer.create(<Roster auth={auth} props={history} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('actually have a table element', () => {
    expect(wrapper.find('table').length).toBe(1);
  });

  describe('call the Auth module (mock) and', () => {
    it('verify user is logged in on componentWillMount', () => {
      const wrap = shallow(<Roster auth={auth} history={history} />);
      expect(auth.loggedIn).toHaveBeenCalled();
      wrap.unmount();
    });

    it('should call auth getPlayers (mock) on componentDidMount and load players into component state', () => {
      // Mock Auth API call to Roster and return Promised playerList
      expect(UnwrappedShallowComponent).toBeDefined();
      expect(UnwrappedShallowComponent.length).toBe(1);
      expect(auth.getPlayers).toHaveBeenCalled();

      // Expect Roster state to be loaded with the player list
      // Fist need to "camelize" the API return
      const camelizedPlayers = playerList.players.map(p =>
        ({
          firstName: p.first_name,
          lastName: p.last_name,
          id: p.id,
          rating: p.rating,
          handedness: p.handedness,
        }));

      expect(UnwrappedShallowComponent.state('players').sort()).toEqual(camelizedPlayers.sort());
    });
  });
});
