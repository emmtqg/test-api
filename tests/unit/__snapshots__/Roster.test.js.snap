// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Roster Table should render correctly 1`] = `
<main
  className="Roster-main-2"
>
  <div
    id="Roster"
  >
    <div
      id="Dialog"
    />
    <div
      classes="Roster-alignTop-8"
    >
      <div
        className="Roster-col-9"
      >
        <h5
          className="MuiTypography-root-27 MuiTypography-h5-43"
        >
          <em>
            The
          </em>
           Roster
        </h5>
      </div>
      <div
        className="Roster-col-9"
      >
        <button
          aria-label="Add"
          className="MuiButtonBase-root-73 MuiFab-root-63 MuiFab-sizeSmall-71 Roster-fab-6"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiFab-label-64"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-76"
              focusable="false"
              onClick={[Function]}
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-132"
          />
        </button>
      </div>
    </div>
    <h6
      className="MuiTypography-root-27 MuiTypography-h6-44 Roster-subTitle-10"
    >
      
    </h6>
    <div
      className="MuiPaper-root-85 MuiPaper-elevation2-89 MuiPaper-rounded-86 Roster-root-1"
    >
      <table
        className="MuiTable-root-112 Roster-table-4"
      >
        <thead
          className="MuiTableHead-root-113"
        >
          <tr
            className="MuiTableRow-root-114 MuiTableRow-head-117"
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root-119 MuiTableCell-head-120 Roster-tableCell-5"
              scope="col"
            >
              Player
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root-119 MuiTableCell-head-120 MuiTableCell-alignRight-129 Roster-tableCell-5"
              scope="col"
            >
              Rating
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root-119 MuiTableCell-head-120 MuiTableCell-alignRight-129 Roster-tableCell-5"
              scope="col"
            >
              Preferred Hand
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root-119 MuiTableCell-head-120 MuiTableCell-alignCenter-128 Roster-tableCell-5"
              scope="col"
            >
              Remove
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root-131"
        />
      </table>
    </div>
  </div>
</main>
`;
